File: src/test/java/cz/vhromada/catalog/validator/GameValidatorTest.java
package cz.vhromada.catalog.validator;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.SoftAssertions.assertSoftly;
import static org.mockito.Mockito.verifyZeroInteractions;

import java.util.Collections;

import cz.vhromada.catalog.entity.Game;
import cz.vhromada.catalog.utils.GameUtils;
import cz.vhromada.common.Movable;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.validator.MovableValidatorTest;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.common.validator.ValidationType;
import cz.vhromada.result.Event;
import cz.vhromada.result.Result;
import cz.vhromada.result.Severity;
import cz.vhromada.result.Status;

import org.junit.jupiter.api.Test;

/**
 * A class represents test for class {@link GameValidator}.
 *
 * @author Vladimir Hromada
 */
class GameValidatorTest extends MovableValidatorTest<Game, cz.vhromada.catalog.domain.Game> {

    /**
     * Test method for {@link GameValidator#GameValidator(MovableService)} with null service for games.
     */
    @Test
    void constructor_NullGameService() {
        assertThatThrownBy(() -> new GameValidator(null)).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null name.
     */
    @Test
    void validate_Deep_NullName() {
        final Game game = getValidatingData(1);
        game.setName(null);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_NAME_NULL", "Name mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with empty name.
     */
    @Test
    void validate_Deep_EmptyName() {
        final Game game = getValidatingData(1);
        game.setName("");

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_NAME_EMPTY", "Name mustn't be empty string.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null URL to english
     * Wikipedia page about game.
     */
    @Test
    void validate_Deep_NullWikiEn() {
        final Game game = getValidatingData(1);
        game.setWikiEn(null);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_WIKI_EN_NULL",
                "URL to english Wikipedia page about game mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null URL to czech
     * Wikipedia page about game.
     */
    @Test
    void validate_Deep_NullWikiCz() {
        final Game game = getValidatingData(1);
        game.setWikiCz(null);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents()).isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_WIKI_CZ_NULL",
                "URL to czech Wikipedia page about game mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with not positive
     * count of media.
     */
    @Test
    void validate_Deep_NotPositiveMediaCount() {
        final Game game = getValidatingData(1);
        game.setMediaCount(0);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null other data.
     */
    @Test
    void validate_Deep_NullOtherData() {
        final Game game = getValidatingData(1);
        game.setOtherData(null);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_OTHER_DATA_NULL", "Other data mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    /**
     * Test method for {@link GameValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null note.
     */
    @Test
    void validate_Deep_NullNote() {
        final Game game = getValidatingData(1);
        game.setNote(null);

        final Result<Void> result = getMovableValidator().validate(game, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "GAME_NOTE_NULL", "Note mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    @Override
    protected MovableValidator<Game> getMovableValidator() {
        return new GameValidator(getMovableService());
    }

    @Override
    protected Game getValidatingData(final Integer id) {
        return GameUtils.newGame(id);
    }

    @Override
    protected cz.vhromada.catalog.domain.Game getRepositoryData(final Game validatingData) {
        return GameUtils.newGameDomain(validatingData.getId());
    }

    @Override
    protected cz.vhromada.catalog.domain.Game getItem1() {
        return GameUtils.newGameDomain(1);
    }

    @Override
    protected cz.vhromada.catalog.domain.Game getItem2() {
        return GameUtils.newGameDomain(2);
    }

    @Override
    protected String getName() {
        return "Game";
    }

}

File: src/test/kotlin/cz/vhromada/catalog/validator/GameValidatorTest.kt
package cz.vhromada.catalog.validator

import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import cz.vhromada.catalog.entity.Game
import cz.vhromada.catalog.utils.GameUtils
import cz.vhromada.common.test.validator.MovableValidatorTest
import cz.vhromada.common.validator.AbstractMovableValidator
import cz.vhromada.common.validator.MovableValidator
import cz.vhromada.common.validator.ValidationType
import cz.vhromada.validation.result.Event
import cz.vhromada.validation.result.Severity
import cz.vhromada.validation.result.Status
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test

/**
 * A class represents test for class [GameValidator].
 *
 * @author Vladimir Hromada
 */
class GameValidatorTest : MovableValidatorTest<Game, cz.vhromada.catalog.domain.Game>() {

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null name.
     */
    @Test
    fun validate_Deep_NullName() {
        val game = getValidatingData(1)
                .copy(name = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_NAME_NULL", "Name mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with empty name.
     */
    @Test
    fun validate_Deep_EmptyName() {
        val game = getValidatingData(1)
                .copy(name = "")

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_NAME_EMPTY", "Name mustn't be empty string.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null URL to english Wikipedia page about game.
     */
    @Test
    fun validate_Deep_NullWikiEn() {
        val game = getValidatingData(1)
                .copy(wikiEn = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_WIKI_EN_NULL", "URL to english Wikipedia page about game mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null URL to czech Wikipedia page about game.
     */
    @Test
    fun validate_Deep_NullWikiCz() {
        val game = getValidatingData(1)
                .copy(wikiCz = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_WIKI_CZ_NULL",
                    "URL to czech Wikipedia page about game mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null count of media.
     */
    @Test
    fun validate_Deep_NullMediaCount() {
        val game = getValidatingData(1)
                .copy(mediaCount = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_MEDIA_COUNT_NULL", "Count of media mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with not positive count of media.
     */
    @Test
    fun validate_Deep_NotPositiveMediaCount() {
        val game = getValidatingData(1)
                .copy(mediaCount = 0)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_MEDIA_COUNT_NOT_POSITIVE", "Count of media must be positive number.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null other data.
     */
    @Test
    fun validate_Deep_NullOtherData() {
        val game = getValidatingData(1)
                .copy(otherData = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_OTHER_DATA_NULL", "Other data mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null note.
     */
    @Test
    fun validate_Deep_NullNote() {
        val game = getValidatingData(1)
                .copy(note = null)

        val result = getValidator().validate(game, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "GAME_NOTE_NULL", "Note mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    override fun getValidator(): MovableValidator<Game> {
        return GameValidator(service)
    }

    override fun getValidatingData(id: Int?): Game {
        return GameUtils.newGame(id)
    }

    override fun getValidatingData(id: Int?, position: Int?): Game {
        return GameUtils.newGame(id)
                .copy(position = position)
    }

    override fun getRepositoryData(validatingData: Game): cz.vhromada.catalog.domain.Game {
        return GameUtils.newGameDomain(validatingData.id)
    }

    override fun getItem1(): cz.vhromada.catalog.domain.Game {
        return GameUtils.newGameDomain(1)
    }

    override fun getItem2(): cz.vhromada.catalog.domain.Game {
        return GameUtils.newGameDomain(2)
    }

    override fun getName(): String {
        return "Game"
    }

}